### PR TITLE
qemu.py fixes

### PR DESCRIPTION
**Pull request by apahim**
@clebergnu, @ldoktor, considering the feedback I got here:
https://lists.gnu.org/archive/html/qemu-devel/2017-07/msg06476.html

Could you please review this PR? Last commit lacks commit message, but that will come later.

Thanks!

### CODE REVIEW

File: scripts/qemu.py [hunk 1]
```diff
@@ -125,38 +134,46 @@ def _post_shutdown(self):
...
+
+        start = time.time()
+        end = start + timeout
+        while time.time() < end:
+            time.sleep(0.1)
+            if not self.is_running():
+                self._load_io_log()
+                raise QEMULaunchError('Error launching QEMU.\n'
+                                      'Command: \n%s\nOutput: \n%s\n' %
+                                      (' '.join(args), ''.join(self._iolog)))
```
Comment (ldoktor) [inline scripts/qemu.py:161]: I don't really like the `timeout` as it's not a timeout, but more like `is_running_timeslot`. Also by default it fails with qemu executions which are meant to finish immediately (you can avoid that by manually setting `launch(-1)`, but it changes the default) and it makes every test if 0.1s (or timeout) longer even though you probably find out the machine is not running on interaction (monitor command, ...).

Also it's unlikely that the process would die and re-execute so you can get (more-less) the same behavior (for non-malfunctioning execution) by simple `time.sleep(time.time() + timeout); self.is_running()`.

If you really want feature like this I'd support `watch_period` argument to `is_running` and allow users to call it from tests:

```python
vm.launch()
if not vm.is_running(10, 0.1):
    ...
```

File: scripts/qemu.py [hunk 2]
```diff
@@ -99,8 +105,11 @@ def get_pid(self):
         return self._popen.pid
 
     def _load_io_log(self):
-        with open(self._qemu_log_path, "r") as fh:
-            self._iolog = fh.read()
+        try:
+            with open(self._qemu_log_path, "r") as fh:
+                self._iolog = fh.read()
+        except:
```
Comment (ldoktor) [inline scripts/qemu.py:111]: I'd recommend `except IOError`.

File: scripts/qemu.py (hunk 1 above)
Comment (apahim) [reply scripts/qemu.py:161]: Yes... I was not sure about this... I will remove it.

File: scripts/qemu.py (hunk 2 above)
Comment (apahim) [reply scripts/qemu.py:111]: ack

File: scripts/qemu.py [hunk 3]
```diff
@@ -86,12 +86,12 @@ def _remove_if_exists(path):
             raise
 
     def is_running(self):
-        return self._popen and (self._popen.returncode is None)
+        return self._popen and (self._popen.poll() is None)
```
Comment (ldoktor) [inline scripts/qemu.py:95]: This includes the regression I mentioned in email:

```python
    >>> import qemu, os
    >>> a=qemu.QEMUMachine("/usr/bin/qemu-kvm", debug=True)
    >>> a.launch()
    QMP:>>> {'execute': 'qmp_capabilities'}
    QMP:<<< {u'return': {}}
    >>> os.kill(a.get_pid(), 9)
    >>> a.is_running()
    False
    >>> a.shutdown()
    >>> os.path.exists(a._qemu_log_path)
    True
```

It is addressed by the last commit so I suggest just reorganizing those patches.

**Review by ldoktor — CHANGES_REQUESTED** [review]
Looks almost good, please re-arrange the commits and address the minor issues in last commit and I think it's ready for upstream.

File: scripts/qemu.py (hunk 3 above)
Comment (apahim) [reply scripts/qemu.py:95]: ok, sure

**Comment by apahim** [comment]
Thank you @ldoktor

**Pull request closed without merging**